File: backend/api/greader_types.go
package api

type greaderTokenType struct {
	Username string `json:"username"`
	Token    string `json:"token"`
}

type greaderHandlelogin struct {
	SID  string `json:"SID"`
	LSID string `json:"LSID"`
	Auth string `json:"Auth"`
}

type greaderUserInfo struct {
	UserId   string `json:"userId"`
	Username string `json:"userName"`
}

type greaderCategory struct {
	Id    string `json:"id"`
	Label string `json:"label"`
}

type greaderSubscription struct {
	Title         string            `json:"title"`
	FirstItemMsec string            `json:"firstitemmsec"`
	HtmlUrl       string            `json:"htmlUrl"`
	IconUrl       string            `json:"iconUrl"`
	SortId        string            `json:"sortid"`
	Id            string            `json:"id"`
	Categories    []greaderCategory `json:"categories"`
}

type greaderSubscriptionList struct {
	Subscriptions []greaderSubscription `json:"subscriptions"`
}

type greaderCanonical struct {
	Href string `json:"href"`
}

type greaderContent struct {
	Direction string `json:"direction,omitempty"`
	Content   string `json:"content"`
}

type greaderOrigin struct {
	StreamId string `json:"streamId"`
	Title    string `json:"title"`
	HtmlUrl  string `json:"htmlUrl"`
}

type greaderItemContent struct {
	CrawlTimeMsec string             `json:"crawlTimeMsec"`
	TimestampUsec string             `json:"timestampUsec"`
	Id            string             `json:"id"`
	Categories    []string           `json:"categories"`
	Title         string             `json:"title"`
	Published     int64              `json:"published"`
	Canonical     []greaderCanonical `json:"canonical"`
	Alternate     []greaderCanonical `json:"alternate"`
	Summary       greaderContent     `json:"summary"`
	Origin        greaderOrigin      `json:"origin"`
}

type greaderItemRef struct {
	Id              string   `json:"id"`
	DirectStreamIds []string `json:"directStreamIds"`
	TimestampUsec   string   `json:"timestampUsec"`
}

type greaderStreamItemIds struct {
	Items    []greaderItemContent `json:"items"`
	ItemRefs []greaderItemRef     `json:"itemRefs"`
}

type greaderStreamItemsContents struct {
	Id      string               `json:"id"`
	Updated int64                `json:"updated"`
	Items   []greaderItemContent `json:"items"`
}

File: frontend/src/api/greader_types.ts
export const enum GReaderURI {
  Login = '/greader/reader/api/0/login',
  Token = '/greader/reader/api/0/token',
  EditTag = '/greader/reader/api/0/edit-tag',
  SubscriptionList = '/greader/reader/api/0/subscription/list',
  StreamItemIds = '/greader/reader/api/0/stream/items/ids',
  StreamItemContents = '/greader/reader/api/0/stream/items/contents',
  MarkAllAsRead = '/greader/reader/api/0/mark-all-as-read',
}

export const enum GReaderTag {
  MarkRead = 'user/-/state/com.google/read',
}

export const enum GReaderStream {
  ReadingList = 'user/-/state/com.google/reading-list',
}

/**
 * The following several interfaces conform to GReader API responses.
 */
export interface GReaderHandleLogin {
  SID: string;
  LSID: string;
  Auth: string;
}

interface GReaderCategory {
  id: string;
  label: string;
}

export interface GReaderSubscription {
  title: string;
  firstItemMsec: string;
  htmlUrl: string;
  iconUrl: string;
  sortId: string;
  id: string;
  categories: GReaderCategory[];
}

export interface GReaderSubscriptionList {
  subscriptions: GReaderSubscription[];
}

interface GReaderCanonical {
  href: string;
}

interface GReaderContent {
  direction: string;
  content: string;
}

interface GReaderOrigin {
  streamId: string;
  title: string;
  htmlUrl: string;
}

export interface GReaderItemContent {
  crawlTimeMsec: string;
  timestampUsec: string;
  id: string;
  categories: string[];
  title: string;
  published: number;
  canonical: GReaderCanonical[];
  alternate: GReaderCanonical[];
  summary: GReaderContent;
  origin: GReaderOrigin;
}

export interface GReaderItemRef {
  id: string;
  directStreamIds: string[];
  timestampUsec: string;
}

export interface GReaderStreamIds {
  itemRefs: GReaderItemRef[];
  continuation: string;
}

export interface GReaderStreamContents {
  id: string;
  updated: number;
  items: GReaderItemContent[];
}